src/subscription_manager/cli_command/register.py

```diff
@@ -329,39 +327,45 @@ def _prompt_for_environment(self):
 
-        if not supports_environments and self.options.environments is not None:
-            system_exit(os.EX_UNAVAILABLE, _("Error: Server does not support environments."))
+        if not supports_environments:
+            if self.options.environments is not None:
+                system_exit(os.EX_UNAVAILABLE, _("Error: Server does not support environments."))
+            return None
+
+        if self.options.environments:
+            environments = self.options.environments
+        else:
+            # We have an activation key, so don't need to fill/check the bits
+            # related to environments, as they are part of the activation key
+            if self.options.activation_keys:
+                return None
 
-        if supports_environments:
             all_env_list = admin_cp.getEnvironmentList(owner_key)
-            if self.options.environments:
-                environments = options.environments
-            else:
-                # If there aren't any environments, don't prompt for one
-                if not all_env_list:
-                    return None
-
-                # If the envronment list is len 1, pick that environment
-                if len(all_env_list) == 1:
-                    log.debug("Using the only available environment: \"{name}\"".format(name=all_env_list[0]["name"]))
-                    return all_env_list[0]['id']
-
-                env_name_list = [env['name'] for env in all_env_list]
-                print(_('Hint: Organization "{key}" contains following environments: {list}').format(
-                      key=owner_key, list=", ".join(env_name_list)))
-                environments = self._prompt_for_environment()
-                if not self.cp.has_capability(MULTI_ENV) and ',' in environments:
-                    system_exit(os.EX_USAGE, _("The entitlement server does not allow multiple environments"))
-
-            return check_set_environment_names(all_env_list, environments)
+            # If there aren't any environments, don't prompt for one
+            if not all_env_list:
+                return None
+
+            # If the envronment list is len 1, pick that environment
+            if len(all_env_list) == 1:
+                log.debug("Using the only available environment: \"{name}\"".format(name=all_env_list[0]["name"]))
+                return all_env_list[0]['id']
+
+            env_name_list = [env['name'] for env in all_env_list]
+            print(_('Hint: Organization "{key}" contains following environments: {list}').format(
+                  key=owner_key, list=", ".join(env_name_list)))
+            environments = self._prompt_for_environment()
+            if not self.cp.has_capability(MULTI_ENV) and ',' in environments:
+                system_exit(os.EX_USAGE, _("The entitlement server does not allow multiple environments"))
+
+        return check_set_environment_names(all_env_list, environments)
```

Comment:
all_env_list can be referenced before assignment if at least one environment is specified (along with the other necessary parameters for registration):
```subscription-manager register --serverurl SOME_SATELLITE/rhsm --environments THING --activationkey ABCD --org default --insecure```

```
local variable 'all_env_list' referenced before assignment
```

I'd suggest that check for duplicates separately from checking for environments which don't exist on the server. IMO we should only get the list of all environments when one is needed but none are specified

Reply:
It isn't just removing duplicates, it is also turning environment names into ids :-/